### DISCUSSION TITLE
Error attempting to supply IPOPT with constraint Jacobian using Casadi (C++)

### DISCUSSION THREAD

**Discussion by gbehrendt** [casadi/casadi#3452] [Q&A]
I have coded NMPC using Casadi in C++ and it works, but it is slow. The longest time spent by IPOPT is evaluating the constraint jacobian. In an attempt to speed up the optimization process I tried supplying IPOPT with a Casadi function to compute the jacobian, but it returns the following error. I have attached the error, the code snip where there error occurs, problem info, and the timing statistics I referred to.

I am already doing a multiple shooting method with an RK4 integrator function and I use warm starting for each MPC loop. I am not sure why InitializeIterates and CheckConvergence take so long.

My ultimate goal is to have the optimization process run on the scale of milliseconds so I would appreciate any help on how to speed up my optimization using casadi in C++, as well as correcting this error. Thank you!

![Screenshot from 2023-11-06 17-30-40](https://github.com/casadi/casadi/assets/55597372/8e9f74d5-d588-42be-a694-8163a527485c)
![Screenshot from 2023-11-06 17-32-45](https://github.com/casadi/casadi/assets/55597372/060b4d8b-16cb-4f6e-9283-baae4d18c9e5)
![Screenshot from 2023-11-07 11-12-43](https://github.com/casadi/casadi/assets/55597372/f61535cd-d365-41a5-aff0-3e60319822a3)
![Screenshot from 2023-11-07 11-11-50](https://github.com/casadi/casadi/assets/55597372/86eab14f-389a-4b35-b878-5aacdbe0c871)

**Comment by jgillis** [comment]
A general rule of thumb to figure out the correct input/output scheme of a solver helper Function: create a standard solver, and query the default helper functions: solver.get_function('nlp_jac_'). You'll find that you need to add a second inout argument for parameters (which may well be of size 0-by-1)

**Reply by gbehrendt** [reply]
Thank you for your response! I was able to correct the error I was getting. What is the correct way to give ipopt the jacobian function I created with casadi's "jacobian(gg,V)" so that ipopt does not need to evaluate the jacobian itself? My suspicion is that the function evaluation of the Equality constraint jacobian takes the majority of the time because ipopt is attempting evaluate that expression itself instead of using my pre-computed jacobian with casadi.

**Comment by dhathris** [comment]
@jgillis @gbehrendt this Q&A has been very useful to me. I think the custom Jacobian expects 2 inputs(design-vector, parameters) and 2 outputs (constraints and jacobian of the constraints w.r.t. design_vector). My callback conforms to this expectation. I ended up composing the constraints of the NLP using the first output from my callback and then I also specify custom jacobian callback function ("jac_g") in the nlp options. However, when I take this approach, IPOPT is taking WAY TOO MUCH memory and my program is crashing. Is this approach reasonable? Or should I configure the constraints expression of the NLP using regular MX variables and let custom Jacobian definition use the callback? What could be the potential reasons for the memory blow up in IPOPT?
